app/src/test/java/org/p2p/wallet/user/interactor/TokenMetadataInteractorTest.kt

```diff
@@ -46,41 +53,76 @@ internal class TokenMetadataInteractorTest {
+        every { gson.fromJson(metadataJson, TokensMetadataInfo::class.java) }.returns(expectedMetadata)
+        coEvery { externalStorageRepository.readJsonFile(any()) }.returns(ExternalFile(metadataJson))
+        coEvery { metadataRepository.loadTokensMetadata(any()) }.returns(UpdateTokenMetadataResult.NoUpdate)
+
+        // WHEN
+        tokenMetadataInteractor.loadAllTokensMetadata()
+
+        // THEN
+        coVerifyOnce { userLocalRepository.setTokenData(expectedMetadata.tokens) }
+        coVerifyNone { externalStorageRepository.saveAsJsonFile(expectedMetadata, any()) }
+    }
+
+    @Test
+    fun `GIVEN Error from remote WHEN loadAllTokensData THEN nothing happens`() = runTest {
+        // GIVEN
+        coEvery { externalStorageRepository.readJsonFile(any()) }.returns(null)
+        coEvery { metadataRepository.loadTokensMetadata(any()) }.returns(UpdateTokenMetadataResult.Error(Throwable()))
+
+        // WHEN
+        tokenMetadataInteractor.loadAllTokensMetadata()
+
+        // THEN
+        coVerifyNone { userLocalRepository.setTokenData(any()) }
+        coVerifyNone { externalStorageRepository.saveAsJsonFile(any<TokensMetadataInfo>(), any()) }
+    }
+
+    @Test
+    fun `GIVEN error from IO THEN method throws and cache is not updated`() = runTest {
+        // GIVEN
+        coEvery { externalStorageRepository.readJsonFile(any()) }.throws(IOException("Illegal access denied"))
+
+        // WHEN
+        val result = kotlin.runCatching { tokenMetadataInteractor.loadAllTokensMetadata() }
+
+        // THEN
+        result.assertThat().matchesPredicate { it.exceptionOrNull() is IOException }
+        coVerifyNone { userLocalRepository.setTokenData(any()) }
+        coVerifyNone { externalStorageRepository.saveAsJsonFile(any<TokensMetadataInfo>(), any()) }
     }
 
     @Test
-    fun `GIVEN metadata WHEN metadata has changes THEN update local file and use a new one`() = runTest {
-        val timestamp = "some timestamp"
-        val file: ExternalFile = mockk()
-        val tokens: List<TokenMetadata> = mockk()
-        val metadata = TokensMetadataInfo(timestamp, tokens)
-        val metadataResult = UpdateTokenMetadataResult.NewMetadata(metadata)
-
-        every { file.data } returns "metadata json"
-        every { externalStorageRepository.readJsonFile(TOKENS_FILE_NAME) } returns file
-        every { gson.fromJson(any<String>(), TokensMetadataInfo::class.java) } returns metadata
-        every { userLocalRepository.setTokenData(tokens) } just Runs
-        every { externalStorageRepository.saveJson(metadata, TOKENS_FILE_NAME) } returns Unit
-        coEvery { metadataRepository.loadTokensMetadata(any()) } returns metadataResult
-
-        tokenMetadataInteractor.loadAllTokensData()
-
-        verify { tokenMetadataInteractor.updateMemoryCacheAndLocalFile(metadataResult) }
+    fun `GIVEN error from gson THEN method throws and cache is not updated`() = runTest {
```

Comment:
you can clearly see that in case of IOException or Gson exception - there will be an exception, so be alert about that